text/0000-or-patterns.md

```diff
@@ -0,0 +1,749 @@
+   `c(p, ..rest) | c(q, ..rest)`.
+
+## Implementation notes
+
+With respect to both static and dynamic semantics,
+it is always valid to first desugar a pattern `c(p | q)`
+in CNF to its equivalent form in DNF, i.e. `c(p) | c(q)`.
+
+However, implementing `c(p | q)` in terms of a pure desugaring to `c(p) | c(q)`
+may not be optimal as the desugaring can result in multiplicative blow-up of patterns.
+An example of such blow up can be seen with:
+
+```rust
+match expr {
+    (0 | 1, 0 | 1, 0 | 1, 0 | 1) => { ... },
+}
+```
+
+If we expanded this naively to DNF we would get:
+
+```rust
+match expr {
+    | (0, 0, 0, 0)
+    | (0, 0, 0, 1)
+    | (0, 0, 1, 0)
+    | (0, 0, 1, 1)
+    | (0, 1, 0, 0)
+    | (0, 1, 0, 1)
+    | (0, 1, 1, 0)
+    | (0, 1, 1, 1)
+    | (1, 0, 0, 0)
+    | (1, 0, 0, 1)
+    | (1, 0, 1, 0)
+    | (1, 0, 1, 1)
+    | (1, 1, 0, 0)
+    | (1, 1, 0, 1)
+    | (1, 1, 1, 0)
+    | (1, 1, 1, 1)
+    => { ... },
+}
+```
+
+Instead, it is more likely that a one-step case analysis will be more efficient.
+
+Which implementation technique to use is left open to each Rust compiler.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+1. Some parsers will have to be rewritten by a tiny bit;
+   We do this with any syntactic change in the language so
+   there should not be any problem.
+
+# Rationale and alternatives
+[alternatives]: #rationale-and-alternatives
+
+As for why the change as proposed in this RFC should be done,
+it is discussed in the [motivation].
+
+Since we already use `|` for alternation at the top level, the only consistent
+operator syntax for alternations in nested patterns would be `|`.
+Therefore, there are not many design choices to make with respect to *how*
+this change should be done rather than *if*.
+
+With respect to the precedence of `|`, we cannot interpret `i @ p | q`
+as `i @ (p | q)` because it is already legal to write `i @ p | j @ q`
+at the top level of a pattern. Therefore, if we say that `|` binds more tightly,
+then `i @ p | j @ q` will associate as `i @ (p | j @ q)` which as a different
+meaning than what we currently have, thus causing a breaking change.
+
+And even if we could associate `i @ p | q` as `i @ (p | q)` there is a good
+reason why we should not. Simply put, we should understand `@` as a
+pattern / set intersection operator and the operator `|` as the union operator.
+This is analogous to multiplication and addition as well as conjunction and
+disjunction in logic. In these fields, it is customary for multiplication and
+conjunction to bind more tightly. That is, we interpret `a * b + c` as
+`(a * b) + c` and not `a * (b + c)`. Similarly, we interpret `p ∧ q ∨ r`
+as `(p ∧ q) ∨ r` and not `p ∧ (q ∨ r)`.
+
+The only real choice that we do have to make is whether the new addition to the
+pattern grammar should be `pat : .. | pat "|" pat ;` or if it instead should be
+`pat : .. | "|"? pat "|" pat ;`. We have chosen the former for 4 reasons:
```

Comment:
nit: `pat ::= "|"? pat "|" pat;` is definitely bad, since it would allow patterns like `| | A | B | C` via the interpretation `| (| A | B) | C`.  So I'm absolutely in favour of the current plan.